File: 125-valid-palindrome/valid-palindrome.py
# -*- coding:utf-8 -*-


# Given a string, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.
#
# Note:&nbsp;For the purpose of this problem, we define empty string as valid palindrome.
#
# Example 1:
#
#
# Input: &quot;A man, a plan, a canal: Panama&quot;
# Output: true
#
#
# Example 2:
#
#
# Input: &quot;race a car&quot;
# Output: false
#
#


class Solution(object):
    def isPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        s = "".join([c.lower() for c in s if c.isalnum()])
        
        return s == s[::-1]
        

File: solutions/125-valid-palindrome/valid-palindrome.cpp
// Given a string, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.
//
// Note: For the purpose of this problem, we define empty string as valid palindrome.
//
// Example 1:
//
//
// Input: "A man, a plan, a canal: Panama"
// Output: true
//
//
// Example 2:
//
//
// Input: "race a car"
// Output: false
//
//


class Solution {
public:
    bool isPalindrome(string s) {
        int n = s.size();
        int i = 0, j = n - 1;
        while (i < j) {
            if (!isalnum(s[i])) ++i;
            else if (!isalnum(s[j])) --j;
            else {
                if (toupper(s[i]) != toupper(s[j])) return false;
                ++i;
                --j;
            }
        }
        return true;
    }
};
